File: ios-app/PlayoffChallenge/Services/Logger.swift
import Foundation

/// Signal-only logging system for production code.
/// Anomalies only: contract violations, unexpected states, mode switches.
/// NO: payloads, counts, success messages, flow logging.
enum Log {
    case error(String)
    case info(String, scope: String)

    func emit() {
        switch self {
        case .error(let message):
            print(message)
        case .info(let message, let scope):
            print("[\(scope)] \(message)")
        }
    }
}

extension Log {
    static func error(_ message: String) {
        Log.error(message).emit()
    }

    static func info(_ message: String, scope: String) {
        Log.info(message, scope: scope).emit()
    }
}

File: backend/services/logger.js
/**
 * Signal-only logging control
 *
 * Allowed signals:
 * - Contract violations (missing required fields)
 * - Unexpected state (invariant violations)
 * - Mode switches (if decision point matters operationally)
 *
 * Forbidden:
 * - Loop logging
 * - Payload logging
 * - Success confirmations
 * - Count logging (unless actively debugging a bug)
 * - Flow descriptions
 */

const LOG = {
  enabled: true,
  errorOnly: true,  // flip to false ONLY if actively debugging a production issue
};

function logError(message, metadata = null) {
  if (LOG.enabled) {
    if (metadata) {
      console.error(message, metadata);
    } else {
      console.error(message);
    }
  }
}

function logWarn(message, metadata = null) {
  if (LOG.enabled && !LOG.errorOnly) {
    if (metadata) {
      console.warn(message, metadata);
    } else {
      console.warn(message);
    }
  }
}

module.exports = {
  LOG,
  logError,
  logWarn
};
